server/container_list.go

```diff
@@ -28,6 +28,38 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
 	return true
 }
 
+// filterContainerList applies a protobuf-defined filter to retrieve only intended containers. Not matching
+// the filter is not considered an error but will return an empty response.
+func (s *Server) filterContainerList(filter *pb.ContainerFilter, origCtrList []*oci.Container) []*oci.Container {
+	// Filter using container id and pod id first.
+	if filter.Id != "" {
+		id, err := s.CtrIDIndex().Get(filter.Id)
+		if err != nil {
+			// If we don't find a container ID with a filter, it should not
+			// be considered an error.  Log a warning and return an empty struct
+			logrus.Warn("unable to find container ID %s", filter.Id)
+			return []*oci.Container{}
+		}
+		c := s.ContainerServer.GetContainer(id)
+		if c != nil {
+			if filter.PodSandboxId == "" || c.Sandbox() == filter.PodSandboxId {
+				return []*oci.Container{}
+			}
+			return []*oci.Container{c}
+		}
+	} else {
+		if filter.PodSandboxId != "" {
+			pod := s.ContainerServer.GetSandbox(filter.PodSandboxId)
+			if pod == nil {
+				return []*oci.Container{}
+			} else {
+				return pod.Containers().List()
+			}
+		}
+	}
+	return origCtrList
```

Comment:
It probably would be good to put up a debug statement before this return saying a filter list couldn't be determined or some such.